rubix-core/src/main/java/com/qubole/rubix/core/RemoteReadRequestChain.java

```diff
@@ -61,78 +79,63 @@ public Integer call()
-                    totalSuffixRead += readAndCopy(suffixBuffer, 0, mbuf, suffixBufferLength);
-                    log.debug(String.format("Read %d bytes into suffix buffer", suffixBufferLength));
-                }
-            }
-        }
-        finally {
-            if (fc != null) {
-                fc.close();
-            }
-            if (localFile != null) {
-                localFile.close();
+        for (ReadRequest readRequest : readRequests) {
+            //560988160, 561671821, 561655437, 561671821, 0]
+            log.debug(String.format("Executing ReadRequest: [%d, %d, %d, %d, %d]", readRequest.getBackendReadStart(), readRequest.getBackendReadEnd(), readRequest.getActualReadStart(), readRequest.getActualReadEnd(), readRequest.getDestBufferOffset()));
+            inputStream.seek(readRequest.actualReadStart);
+            log.debug(String.format("Trying to Read %d bytes into destination buffer", readRequest.getActualReadLength()));
+            int readBytes = readIntoDestinationBuffer(readRequest.getDestBuffer(), readRequest.destBufferOffset, readRequest.getActualReadLength());
+            log.debug(String.format("Read %d bytes into destination buffer", readBytes));
+            totalRequestedRead += readBytes;
+
+            long cacheBeginBlock = getCacheBeginBlock(readRequest);
+            long cacheEndBlock = getCacheEndBlock(readRequest);
+            if (cacheBeginBlock < cacheEndBlock) {
+                long cacheReadStart = cacheBeginBlock * blockSize;
+                int prefixOffset = (int) (cacheReadStart - readRequest.actualReadStart);
+                int suffixOffset = (int) (readRequest.actualReadEnd - cacheEndBlock * blockSize);
+                log.debug(String.format("Cache begin block- %d, end block- %d, prefix offset- %d, suffix offset- %d",
+                        cacheBeginBlock, cacheEndBlock, prefixOffset, suffixOffset));
+                int cachedBytes = copyIntoCache(readRequest.destBuffer, readRequest.destBufferOffset + prefixOffset, readRequest.getActualReadLength() - suffixOffset - prefixOffset, cacheReadStart);
+                log.debug(String.format("Copied %d bytes into cache", cachedBytes));
             }
         }
-        log.info(String.format("Read %d bytes from remote file, added %d to destination buffer", totalPrefixRead + totalRequestedRead + totalSuffixRead, totalRequestedRead));
         return totalRequestedRead;
     }
 
-    private int readAndCopy(byte[] destBuffer, int destBufferOffset, MappedByteBuffer localFileBuffer, int length)
+    // buffer, 0, 16384
+    private int readIntoDestinationBuffer(byte[] destBuffer, int destBufferOffset, int length)
             throws IOException
     {
         int nread = 0;
         while (nread < length) {
             int nbytes = inputStream.read(destBuffer, destBufferOffset + nread, length - nread);
             if (nbytes < 0) {
-                return nread;
+                break;
```

Comment:
this should be `return` not `break`, keep it how it was before else we can get into errors